<a href="https://colab.research.google.com/github/lakshmant-thota/ml_zoomcamp/blob/master/hw3_mlzp_lakshman.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
!wget https://archive.ics.uci.edu/static/public/222/bank+marketing.zip

--2024-10-13 08:19:44--  https://archive.ics.uci.edu/static/public/222/bank+marketing.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified
Saving to: ‘bank+marketing.zip’

bank+marketing.zip      [  <=>               ] 999.85K  2.42MB/s    in 0.4s    

2024-10-13 08:19:45 (2.42 MB/s) - ‘bank+marketing.zip’ saved [1023843]



In [ ]:
df = pd.read_csv('/content/sample_data/bank-full.csv',sep=';')

In [ ]:
df.head()

age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married    unknown      no     1506     yes   no   
4   33       unknown   single    unknown      no        1      no   no   

   contact  day month  duration  campaign  pdays  previous poutcome   y  
0  unknown    5   may       261         1     -1         0  unknown  no  
1  unknown    5   may       151         1     -1         0  unknown  no  
2  unknown    5   may        76         1     -1         0  unknown  no  
3  unknown    5   may        92         1     -1         0  unknown  no  
4  unknown    5   may       198         1     -1         0  unknown  no

In [ ]:
del df['loan']
del df['default']

In [ ]:
df.head()

age           job  marital  education  balance housing  contact  day month  \
0   58    management  married   tertiary     2143     yes  unknown    5   may   
1   44    technician   single  secondary       29     yes  unknown    5   may   
2   33  entrepreneur  married  secondary        2     yes  unknown    5   may   
3   47   blue-collar  married    unknown     1506     yes  unknown    5   may   
4   33       unknown   single    unknown        1      no  unknown    5   may   

   duration  campaign  pdays  previous poutcome   y  
0       261         1     -1         0  unknown  no  
1       151         1     -1         0  unknown  no  
2        76         1     -1         0  unknown  no  
3        92         1     -1         0  unknown  no  
4       198         1     -1         0  unknown  no

In [ ]:
df.size

678165

In [ ]:
df.columns

Index(['age', 'job', 'marital', 'education', 'balance', 'housing', 'contact',
       'day', 'month', 'duration', 'campaign', 'pdays', 'previous', 'poutcome',
       'y'],
      dtype='object')

In [ ]:
df.dtypes

age           int64
job          object
marital      object
education    object
balance       int64
housing      object
contact      object
day           int64
month        object
duration      int64
campaign      int64
pdays         int64
previous      int64
poutcome     object
y            object
dtype: object

In [ ]:
df.describe()

age        balance           day      duration      campaign  \
count  45211.000000   45211.000000  45211.000000  45211.000000  45211.000000   
mean      40.936210    1362.272058     15.806419    258.163080      2.763841   
std       10.618762    3044.765829      8.322476    257.527812      3.098021   
min       18.000000   -8019.000000      1.000000      0.000000      1.000000   
25%       33.000000      72.000000      8.000000    103.000000      1.000000   
50%       39.000000     448.000000     16.000000    180.000000      2.000000   
75%       48.000000    1428.000000     21.000000    319.000000      3.000000   
max       95.000000  102127.000000     31.000000   4918.000000     63.000000   

              pdays      previous  
count  45211.000000  45211.000000  
mean      40.197828      0.580323  
std      100.128746      2.303441  
min       -1.000000      0.000000  
25%       -1.000000      0.000000  
50%       -1.000000      0.000000  
75%       -1.000000      0.000000  
max      871.000000    275.000000

In [ ]:
df.isnull().sum()

age          0
job          0
marital      0
education    0
balance      0
housing      0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
y            0
dtype: int64

In [ ]:
df.shape

(45211, 15)

# Question 1
What is the most frequent observation (mode) for the column education?

unknown
primary
secondary
tertiary

In [ ]:
df.education.mode()

0    secondary
Name: education, dtype: object

In [ ]:
df.education.value_counts()

education
secondary    23202
tertiary     13301
primary       6851
unknown       1857
Name: count, dtype: int64

In [ ]:
numerical = df.select_dtypes(include=['int64','float64']).columns
print(numerical)

Index(['age', 'balance', 'day', 'duration', 'campaign', 'pdays', 'previous'], dtype='object')


In [ ]:
df[numerical]

age  balance  day  duration  campaign  pdays  previous
0       58     2143    5       261         1     -1         0
1       44       29    5       151         1     -1         0
2       33        2    5        76         1     -1         0
3       47     1506    5        92         1     -1         0
4       33        1    5       198         1     -1         0
...    ...      ...  ...       ...       ...    ...       ...
45206   51      825   17       977         3     -1         0
45207   71     1729   17       456         2     -1         0
45208   72     5715   17      1127         5    184         3
45209   57      668   17       508         4     -1         0
45210   37     2971   17       361         2    188        11

[45211 rows x 7 columns]

# Answer 1
sconday is the most repeated value or mode in education

# Question 2
Create the correlation matrix for the numerical features of your dataset. In a correlation matrix, you compute the correlation coefficient between every pair of features.

What are the two features that have the biggest correlation?

age and balance
day and campaign
day and pdays
pdays and previous

In [ ]:
df[numerical].corr(method='pearson')

age   balance       day  duration  campaign     pdays  previous
age       1.000000  0.097783 -0.009120 -0.004648  0.004760 -0.023758  0.001288
balance   0.097783  1.000000  0.004503  0.021560 -0.014578  0.003435  0.016674
day      -0.009120  0.004503  1.000000 -0.030206  0.162490 -0.093044 -0.051710
duration -0.004648  0.021560 -0.030206  1.000000 -0.084570 -0.001565  0.001203
campaign  0.004760 -0.014578  0.162490 -0.084570  1.000000 -0.088628 -0.032855
pdays    -0.023758  0.003435 -0.093044 -0.001565 -0.088628  1.000000  0.454820
previous  0.001288  0.016674 -0.051710  0.001203 -0.032855  0.454820  1.000000

In [ ]:
# prompt: how to extract the highest correlated pair in the above matric or a pair with highest correlation followed by the lower ones

correlation_matrix = df[numerical].corr()

# Remove diagonal part of the matrix
correlation_matrix = correlation_matrix.mask(np.equal(*np.indices(correlation_matrix.shape)))

# Find the highest correlation
max_correlation = np.nanmax(correlation_matrix)
row_max, col_max = correlation_matrix[correlation_matrix == max_correlation].stack().idxmax()

print(f"Highest correlation: {row_max} and {col_max} with correlation {max_correlation}")

# Sort the correlations in descending order and print the pairs
sorted_correlations = correlation_matrix.unstack().sort_values(ascending=False)
for (row, col), correlation in sorted_correlations.items():
    if not np.isnan(correlation):
      print(f"{row} and {col}: {correlation}")


Highest correlation: pdays and previous with correlation 0.4548196354805043
previous and pdays: 0.4548196354805043
pdays and previous: 0.4548196354805043
campaign and day: 0.1624902163261922
day and campaign: 0.1624902163261922
balance and age: 0.09778273937134807
age and balance: 0.09778273937134807
duration and balance: 0.021560380494668893
balance and duration: 0.021560380494668893
previous and balance: 0.016673636728357952
balance and previous: 0.016673636728357952
campaign and age: 0.0047603117624537014
age and campaign: 0.0047603117624537014
day and balance: 0.004502585129718555
balance and day: 0.004502585129718555
balance and pdays: 0.003435321868106611
pdays and balance: 0.003435321868106611
previous and age: 0.0012883192026692438
age and previous: 0.0012883192026692438
previous and duration: 0.0012030569167003798
duration and previous: 0.0012030569167003798
pdays and duration: -0.0015647704813434864
duration and pdays: -0.0015647704813434864
duration and age: -0.0046484284706

# Answer 2
the answer for Q2 is pdays and previous with correlation of 0.4548

#Target encoding
Now we want to encode the y variable.
Let's replace the values yes/no with 1/0.

In [41]:
df.y = df.y.map({'yes':1,'no':0})

In [42]:
df.y.value_counts()

y
0    39922
1     5289
Name: count, dtype: int64

In [43]:
df.y.mean()

0.11698480458295547

#Split the data
Split your data in train/val/test sets with 60%/20%/20% distribution.
Use Scikit-Learn for that (the train_test_split function) and set the seed to 42.
Make sure that the target value y is not in your dataframe.

In [55]:
from sklearn.model_selection import train_test_split

df_full_train,df_test = train_test_split(df,test_size=0.2,random_state = 42)

In [56]:
df_train,df_val = train_test_split(df_full_train,test_size=0.25,random_state = 42)

In [57]:
len(df_train),len(df_val),len(df_test)

(27126, 9042, 9043)

In [50]:
df.shape

(45211, 15)

In [58]:
y_train = df_train.y.values
y_val = df_val.y.values
y_test = df_test.y.values

del df_train['y']
del df_val['y']
del df_test['y']

# Question 3
Calculate the mutual information score between y and other categorical variables in the dataset. Use the training set only.
Round the scores to 2 decimals using round(score, 2).
Which of these variables has the biggest mutual information score?

contact
education
housing
poutcome

In [59]:
from sklearn.metrics import mutual_info_score

In [60]:
categorical = df.select_dtypes(include=['object','bool_']).columns

In [61]:
categorical

Index(['job', 'marital', 'education', 'housing', 'contact', 'month',
       'poutcome'],
      dtype='object')

In [68]:
def mi_score(series):
  return round(mutual_info_score(series,y_train),2)

In [69]:
mi = df_train[categorical].apply(mi_score)
mi.sort_values(ascending=False)

month        0.03
poutcome     0.03
job          0.01
housing      0.01
contact      0.01
marital      0.00
education    0.00
dtype: float64

# answer 3
the answer is poutocome with score of 0.03

# Question 4
Now let's train a logistic regression.
Remember that we have several categorical variables in the dataset. Include them using one-hot encoding.
Fit the model on the training dataset.
To make sure the results are reproducible across different versions of Scikit-Learn, fit the model with these parameters:
model = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)
Calculate the accuracy on the validation dataset and round it to 2 decimal digits.
What accuracy did you get?

0.6
0.7
0.8
0.9

In [71]:
from sklearn.feature_extraction import DictVectorizer

In [72]:
dv = DictVectorizer(sparse=False)

In [74]:
train_dicts = df_train.to_dict(orient='records')

In [75]:
val_dicts = df_val.to_dict(orient='records')

In [81]:
list(dv.fit_transform(train_dicts[0])[0])

[32.0, 1100.0, 1.0, 1.0, 11.0, 67.0, 1.0, 1.0, 1.0, 1.0, 1.0, -1.0, 1.0, 0.0]

In [82]:
x_train = dv.fit_transform(train_dicts)
x_val = dv.fit_transform(val_dicts)

In [83]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)
model.fit(x_train,y_train)

LogisticRegression(max_iter=1000, random_state=42, solver='liblinear')

In [84]:
y_pred = model.predict(x_val)

In [85]:
y_pred

array([0, 0, 0, ..., 0, 0, 0])

In [87]:
actual_score = (y_pred == y_val).mean()
actual_score

0.9009068790090687

# Answer 4 the accuracy is 0.9

#Question 5
Let's find the least useful feature using the feature elimination technique.
Train a model with all these features (using the same parameters as in Q4).
Now exclude each feature from this set and train a model without it. Record the accuracy for each model.
For each feature, calculate the difference between the original accuracy and the accuracy without the feature.
Which of following feature has the smallest difference?

age
balance
marital
previous

In [88]:
feature_wise_scores = []
for c in df_train.columns:
  df_train_copy = df_train.copy()
  df_val_copy = df_val.copy()
  del df_train_copy[c]
  del df_val_copy[c]
  dv = DictVectorizer(sparse=False)
  train_dicts = df_train_copy.to_dict(orient='records')
  val_dicts = df_val_copy.to_dict(orient='records')
  x_train = dv.fit_transform(train_dicts)
  x_val = dv.fit_transform(val_dicts)
  model = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)
  model.fit(x_train,y_train)
  y_pred = model.predict(x_val)
  score = (y_pred == y_val).mean()
  diff = actual_score - score
  feature_wise_scores.append((c,score,diff))
feature_wise_scores.sort(key = lambda x:x[2])
feature_wise_scores

[('age', 0.9013492590134926, -0.00044238000442387015),
 ('day', 0.9013492590134926, -0.00044238000442387015),
 ('job', 0.9011280690112807, -0.0002211900022119906),
 ('housing', 0.9011280690112807, -0.0002211900022119906),
 ('balance', 0.9010174740101747, -0.0001105950011059953),
 ('marital', 0.9009068790090687, 0.0),
 ('education', 0.9009068790090687, 0.0),
 ('pdays', 0.9009068790090687, 0.0),
 ('previous', 0.9009068790090687, 0.0),
 ('contact', 0.900464499004645, 0.00044238000442375913),
 ('campaign', 0.9002433090024331, 0.0006635700066356387),
 ('month', 0.8998009289980093, 0.0011059500110593978),
 ('poutcome', 0.8933864189338642, 0.007520460075204571),
 ('duration', 0.8897367838973679, 0.011170095111700862)]

# Answer 5 : the lwast useful features are age and day and from the options the answer is "age"

# Question 6
Now let's train a regularized logistic regression.
Let's try the following values of the parameter C: [0, 0.01, 0.1, 1, 10].
Train models using all the features as in Q4.
Calculate the accuracy on the validation dataset and round it to 3 decimal digits.
Which of these C leads to the best accuracy on the validation set?

0.01
0.1
1
10
100

In [105]:

df_train_copy = df_train.copy()
df_val_copy = df_val.copy()
dv = DictVectorizer(sparse=False)
train_dicts = df_train_copy.to_dict(orient='records')
val_dicts = df_val_copy.to_dict(orient='records')
x_train = dv.fit_transform(train_dicts)
x_val = dv.fit_transform(val_dicts)
c_scores = []
for c in [0.01 , 0.1 , 1.0 , 10.0,100.0]:
  model = LogisticRegression(solver='liblinear', C=c, max_iter=1000, random_state=42)
  model.fit(x_train,y_train)
  y_pred = model.predict(x_val)
  score = (y_pred == y_val).mean()
  c_scores.append((c,score))
c_scores.sort(key = lambda x:x[1],reverse = True)
c_scores

[(1.0, 0.9009068790090687),
 (10.0, 0.9009068790090687),
 (0.1, 0.9007962840079629),
 (100.0, 0.9006856890068569),
 (0.01, 0.8979208139792081)]

# Answer 6
both 1 and 10 are equally performing on the model but the question was to select the least